SI-6539 Annotation for methods unfit for post-typer ASTs

retronym opened the pull request:
Motivated by the `.value` method in the SBT task-syntax branch,
which should only be called within the context of the argument
to a setting initialization macro.

The facility is akin to a fatal deprecation.

Review by @xeno-by @harrah

retronym committed:
SI-6539 Annotation for methods unfit for post-typer ASTs

Motivated by the `.value` method in the SBT task-syntax branch,
which should only be called within the context of the argument
to a setting initialization macro.

The facility is akin to a fatal deprecation.

xeno-by commented on src/reflect/scala/reflect/macros/compileTimeOnly.scala:
```diff
@@ -0,0 +1,16 @@
+ *
+ * @param  message the error message to print during compilation if a reference remains
+ *                 after type checking
+ * @since  2.10.0
```
2.10.1?

adriaanm replied:
indeed

retronym replied:
done

xeno-by commented on src/compiler/scala/tools/nsc/typechecker/RefChecks.scala:
```diff
@@ -1376,6 +1376,11 @@ abstract class RefChecks extends InfoTransform with scala.reflect.internal.trans
 
+    private def checkCompileTimeOnly(sym: Symbol, pos: Position) = {
+      if (sym.isCompileTimeOnly)
+        unit.error(pos, s"reference to ${sym.fullLocationString} should not survive typechecking: ${sym.compileTimeOnlyMessage.get}")
```
Why not just show a compileTimeOnlyMessage?

harrah replied:
Agree.

retronym replied:
I've changed this, and removed the default argument for the message.

gkossakowski commented on test/files/neg/t6539.check:
```diff
@@ -0,0 +1,10 @@
+Test_2.scala:2: error: reference to method cto in object M should not survive typechecking: cto may only be used as an argument to m
+	M.cto // error
+          ^
```
Seems like positions for those errors are off.

retronym replied:
Oops, used tabs instead of spaces.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/865/

retronym committed:
Refine @compileTimeOnly

 - Don't default the message, and show it exclusively.
 - Fix cut-and-pasto in the @since tag
 - Be tolerant if the annotaion class is missing, as seems to
   have been the case compiling the continuations plugin.
 - s/\t/  / in the test file to show the errors are positioned
   correctly.
 - Use defensive getOrElse

harrah commented:
LGTM.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/866/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1573/

xeno-by commented:
lgtm as well granted we agree that it's a temporary solution subject to being removed after untyped macros are implemented

retronym commented:
I suspect that the untyped-macros approach might still have a problem: the compile time only macro must know all of the potential enclosing macros in which it is valid, which might stop a user of a library from writing a custom macro to replace the built-in enclosing macros. But we'll see when we get there.

xeno-by commented:
Take a look here: http://www.schemeworkshop.org/2011/papers/Barzilay2011.pdf.

By default the compile-time thingie should always produce a compilation error. It's a job of an enclosing macro to wrap the block with a compile-time thingie into a binding which rebinds the compile-time thingie to something useful within the domain.

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1574/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1574/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/867/

retronym commented:
@xeno-by Oh right, makes total sense.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/873/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1577/

paulp commented:
You know you've found your optimal audience when you can fit "thingie, thingie, thingie, something" into two lines and the response is a non-sarcastic "oh right, makes total sense."

som-snytt commented:
I had the same reaction.  I would have tweeted it, but it seems that neither Henry James nor @xeno-by is tweetable.  Both are succinct as possible but >140.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/883/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1590/

retronym commented:
PLS REBUILD ALL

BTW, @xeno-by I'll leave it up to you to use this on `splice`.

xeno-by commented:
Oh right, thanks for reminding

adriaanm commented on src/compiler/scala/tools/nsc/typechecker/RefChecks.scala:
```diff
@@ -1376,6 +1376,16 @@ abstract class RefChecks extends InfoTransform with scala.reflect.internal.trans
+    private def checkCompileTimeOnly(sym: Symbol, pos: Position) = {
+      if (sym.isCompileTimeOnly) {
+        def defaultMsg =
+          """|Reference to ${sym.fullLocationString} should not have survived past type checking,
```
is this missing the leading s for interpolaShun?

retronym commented:
right you are, PLS REBUILD ALL

paulp commented:
I see I'm not the only one in desperate need of that warning. Keep forgetting to implement it when I'm between distractions.

retronym commented:
:cat: https://scala-webapps.epfl.ch/jenkins/job/scala-checkin-manual/667/console

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1625/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/920/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1625/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/920/

paulp commented:
PLS REBUILD ALL

paulp commented:
A prayer to St. Spurious

retronym commented:
> I see I'm not the only one in desperate need of that warning. Keep forgetting to implement it when I'm between distractions.

Here's another candidate warning:

``` scala
"""|abc
   |def
   |"""
```

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/939/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/939/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1646/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1646/

adriaanm commented:
this is not a forward compatible change, so we'll have to move the annotatoin to the compiler jar or revert this

xeno-by commented:
See the last comment at https://issues.scala-lang.org/browse/SI-6539
